### ISSUE TITLE
Filesystem API is too restrictive

### ISSUE THREAD

**Issue by adam-bates** [ggez/ggez#873]
## TL;DR
There is no way to create and work with a `ggez::filesystem::Filesystem` without creating a `ggez::Context` and using helper functions that require that context.

Also, apologies for the long read in your already long list of issues!

###### Only relevant if you want my "reasoning" behind this

Hey @icefoxen!  (Our first meeting was https://github.com/ggez/ggez/pull/869 but this seems more official)

I'm building a top-down 2D RPG with very simple graphics. Most likely 16x16 tiles in some layout that probably won't exceed 300 pixels wide. Though simplistic, I really like being in full control of my games, and being able to optimize them so they'll run on a potato... So I'm finally at a stage where I've decided on the underlying tech-stack, and have started the initial codebase.

I originally looked at [*Amethyst*](https://amethyst.rs) as it was one of the more popular results for a #RealGameEngine in Rust, but it seems they see [*Bevy*](https://bevyengine.org) as a successor. So I spent probably a good 2 weeks prototyping in Bevy, and I've got to say [@cart](https://github.com/cart) knows what he's doing.

Unfortunately, Bevy both isn't ready, and is trying to be everything to everyone; I ended up spending far too much time fighting with the camera, UI, object culling, performance (no batch rendering), etc etc. And thus sadly, had to start looking elsewhere.

 I found GGEZ and loved how easy it is to draw graphics. The graphics API gives you everything you need and then gets out of your way. I also love how unopinionated things seem to be, all the way down to letting the developer implement the game loop using your API.

I have since also tried [*Coffee*](https://github.com/hecrj/coffee), [*SDL2*](https://github.com/Rust-SDL2/rust-sdl2), [*Piston*](https://www.piston.rs), and a few others. But GGEZ has the perfect compromise between low-level and high-level, that I've decided I'll take inspiration from the [GGEZ game template](https://github.com/ggez/game-template) and add in the missing features around the GGEZ implementation (ie. [Specs](https://github.com/amethyst/specs) for ECS, [fern](https://github.com/daboross/fern) for logging, etc).

This was going all good and well, until...

## Description

An early sketch at how I'd like my main function to look is something like:
```rust

mod config;
mod error;
mod filesystem;
mod game;
mod utils;

use error::types;

fn main() -> types::Result {
    // Initialize filesystem
    let mut fs = filesystem::new_filesystem(); // returns a ggez::filesystem::Filesystem

    // Setup logger
    let log_output = match config::log::setup(&mut fs) {
        Err(error) => return error::handle_log_setup_err(error),
        Ok(output) => output,
    };

    // Run game
    game::run_game(fs).or_else(error::handle_game_err)?;

    // Clean up logger
    config::log::clean_up(log_output)
}
```

Notice how I'm creating a Filesystem object before I run any concept of a "game". I think this is reasonable as my logger, which uses the filesystem to log to the user's local data directory in `--release`, should be setup as early as possible; as well as it makes it trivial to reload the game without exiting the application. It also **does not** make sense to initialize a `ggez::Context` before creating the logger because the context contains window and game details that are important to have logging ready for.

**Unfortunately, I've had to branch out to my own custom build of GGEZ because what I'm trying to do is impossible with the current API, for what seems like no good reason.** I've pushed my build to a private repo as I'm making changes specific to my game for optimization, so I can't link anything, but I'll show some of the things I've done:

```rust
// src/lib.rs
...
pub mod vfs; // Wasn't public before, so couldn't access this entire crate
```
```rust
// src/filesystem.rs
...
pub struct Filesystem {
    // Fields weren't public, so you couldn't customize how the Filesystem was created
    pub vfs: vfs::OverlayFS,
    pub resources_path: path::PathBuf,
    pub zip_path: path::PathBuf,
    pub user_config_path: path::PathBuf,
    pub user_data_path: path::PathBuf,
...
impl Filesystem {
    ...
    pub fn new(id: &str, author: &str) -> GameResult<Filesystem> {
        ...
        // Per-user data dir,
        // ~/.local/share/whatever/
        {
            user_data_path = project_dirs.data_local_dir();
            trace!("User-local data path: {:?}", user_data_path);

            // This was read-only ... why?!
            // I'm assuming it's intentional, but I would like to save user logs and save files here ...
            let physfs = vfs::PhysicalFS::new(&user_data_path, false);
            overlay.push_back(Box::new(physfs));
        }
        ...
    }
    ...
    // Wasn't public, so you were FORCED to use a helper method that takes ctx: &mut ggez::Context
    // which just calls ctx.filesystem.create_dir for you!
    // I've done this change on every method in Filesystem
    pub fn create_dir<P: AsRef<path::Path>>(&mut self, path: P) -> GameResult<()> { ... }
    ...
    // Because these are not generic I've modified them to use my own game config struct
    // And made their path configurable, as it was hard-coded to use "config.toml" before
    pub fn read_config(&mut self) -> GameResult<MyGameConfigStruct> { ... }
    pub fn write_config(&mut self, conf: &MyGameConfigStruct) -> GameResult<()> { ... }
```

I read through your posting here: https://wiki.alopex.li/AGuideToRustGameFrameworks2019 and I think we agree on more than we disagree on, and I really admire your work and thoughts. In the post there's a note:
```
Everything ggez offers is in the form of functions, not methods on Context or Image or such. That’s mainly a matter of taste. It’s also how LÖVE does it, which ggez is based on, so I kept the same idiom.
```

I don't have an issue with this. It keeps the API simple and easy to pick up, and lets the developer decide on the complexity they'd like. But I don't understand why you had to make the underlying implementations private? Doesn't that just make the API more opinionated than it needs to be?

**Of course, if this is intentional by design, and you do not wish to change it, close this issue! I am planning on staying with my custom build of GGEZ for optimizations that are specific to my game.**

**Also, seriously, thank you for all of your hard work. I really REALLY appreciate it!**

**Comment by icefoxen** [comment]
Short answer, I like it in theory but probably won't get around to it any time soon.  The interdependence of filesystem and context creation is something that caused a fair amount of hair-pulling until I came up with a more-or-less-workable solution, that of a `ContextBuilder` that allows you to add filesystem paths before calling `build()`.  But as you've noticed, it doesn't work for all cases.

Pulling ggez's filesystem layer out into its own thing has been a wishlist item for a long time, and some initial work has been done on that [here](https://github.com/ggez/ggez/tree/ggraphics/gvfs), but there's lots of fiddling and cleanup that needs doing.  If you're interested, you might consider digging into that and/or cleaning it up for your own purposes.

> But I don't understand why you had to make the underlying implementations private? Doesn't that just make the API more opinionated than it needs to be?

Basically so that people don't a) confuse the public and private parts, b) get lost in the implementation details while trying to learn ggez instead of just using the nice parts, or c) tinker around with internal state and then ask for help when they mess something up.  It is, I confess, somewhat selfish.  Since you seem more-than-usually interested in its internals and uncompromising about doing the Right Thing for your use case, I don't think you should feel bad at all about customizing ggez's codebase for your needs.  :)

**Comment by adam-bates** [comment]
###### Sorry for the *very* late response!

That definitely makes sense. Thanks for the explanation. I've made it work well enough **for my specific needs** on my private fork, but if I revisit and put some work into officially pulling them apart then I'll definitely look into sharing that work. :)

Cheers!
